Refactor Nuget operations out of Publish-PSArtifactUtility function

Benny1007 opened the pull request:
Refactor nuget operations out of Publish-PSArtifactUtility.ps1 to start #335

edyoung commented on src/PowerShellGet/private/functions/New-NugetPackage.ps1:
```diff
@@ -0,0 +1,110 @@
+        $process.StartInfo = $processStartInfo
+        $process.Start() | Out-Null
+        $process.WaitForExit()
+
```
why not remove the tem path here

Benny1007 replied:
ah yep makes sense. done.

edyoung commented on src/PowerShellGet/private/functions/New-NuspecFile.ps1:
```diff
@@ -0,0 +1,125 @@
+    if ($Tags.Length -gt 4000) {
+        $Tags = $Tags.Substring(0, $Tags.LastIndexOf(" "))
+        Write-Warning -Message "Nuspec 'Tag' list exceeded max 4000 characters and was truncated."
+    }
```
PowerShellGallery does not enforce the 4000 character limit and we have modules which exceed it. Probably better to Write-Warning "Tag list exceeded 4000 characters and may not be accepted by some Nuget feeds" without truncating it

Benny1007 replied:
Done.

jzabroski replied:
@Benny1007  The message should be updated to say you're not truncating it.

edyoung commented on src/PowerShellGet/private/functions/Publish-PSArtifactUtility.ps1:
```diff
@@ -372,267 +329,116 @@ function Publish-PSArtifactUtility
-        $null = Microsoft.PowerShell.Management\New-Item -Path $csprojBasePath -ItemType Directory -Force -WhatIf:$false -Confirm:$false
-        $NuspecPath = Microsoft.PowerShell.Management\Join-Path -Path $csprojBasePath -ChildPath "$Name.nuspec"
-        $CsprojFilePath = Microsoft.PowerShell.Management\Join-Path -Path $csprojBasePath -ChildPath "$Name.csproj"
+        $dependencyObject = New-Object -TypeName PSCustomObject -Property $props
```
You could also do $dependencyObject = [PSCustomObject]$props, btw

Benny1007 replied:
not done

jzabroski replied:
I think the style @Benny1007  wrote here is more readable.

edyoung commented on src/PowerShellGet/private/functions/Publish-PSArtifactUtility.ps1:
```diff
@@ -372,267 +329,116 @@ function Publish-PSArtifactUtility
-            $SemanticVersionString = Get-NormalizedVersionString -Version $Version
-            $NupkgPath = Join-PathUtility -Path $NugetPackageRoot -ChildPath "$Name.$($SemanticVersionString).nupkg" -PathType File
-        }
+    # fix#335 refactor to use new New-NuspecFile function.
```
I don't think we need the comment, it won't mean anything to a future reader of the code

Benny1007 replied:
done.

edyoung commented on src/PowerShellGet/private/functions/Publish-PSArtifactUtility.ps1:
```diff
@@ -372,267 +329,116 @@ function Publish-PSArtifactUtility
-            }
+    try {
+        New-NuspecFile $params
+        $NuspecFullName = Join-Path -Path $NugetPackageRoot -ChildPath "$Name.nuspec"
```
You could consider having New-NuspecFile return the path of the file it created? up to you

Benny1007 replied:
I think i'll leave this if that's ok, I'd rather explicitly check the file exists and have that blow up catching the exception than assign the result to $NuspecFullName then test-path in an if.

Benny1007 replied:
actually scratch that i've made that change.

edyoung commented on src/PowerShellGet/private/functions/Publish-PSArtifactUtility.ps1:
```diff
@@ -372,267 +329,116 @@ function Publish-PSArtifactUtility
-            PassThru = $true
+        elseif ($NuGetExePath) {
+            New-NugetPackage -NuspecPath $NuspecFullName -NugetPackageRoot $NugetPackageRoot -NugetExePath $NuGetExePath -Verbose:$VerbosePreference
         }
```
does there need to be an else to cover producing an error when neither nuget or dotnet were found?

Benny1007 replied:
line 69 should populate these script level variables though with Install-NuGetClientBinaries. Not a fan of that tbh though. If those variables aren't set would you prefer to exit gracefully with an error message? Would Install-NuGetClientBinaries not do that? not looked at that code.

edyoung commented on src/PowerShellGet/private/functions/Publish-PSArtifactUtility.ps1:
```diff
@@ -372,267 +329,116 @@ function Publish-PSArtifactUtility
-            $ArgumentList += @('-source', "`"$($Destination.TrimEnd('\'))`"")
-            $ArgumentList += @('-apikey', "`"$NugetApiKey`"")
-            $ArgumentList += '-NonInteractive'
+        $NupkgFullName = Join-Path -Path $NugetPackageRoot -ChildPath "$Name.$Version.nupkg"
```
New-NugetPackage could return name of the package it created

edyoung commented on src/PowerShellGet/private/functions/Publish-PSArtifactUtility.ps1:
```diff
@@ -372,267 +329,116 @@ function Publish-PSArtifactUtility
-            if($errorMsg) {
-                Write-Verbose -Message $errorMsg
-            }
+    # Fix#335 refactor to use new Publish-NugetPackage function.
```
I'm not seeing Publish-NugetPackage in this PR?

Benny1007 replied:
Oh wow yeah, have re-added.

Benny1007 commented on src/PowerShellGet/private/functions/New-NugetPackage.ps1:
```diff
@@ -0,0 +1,110 @@
+
+    if ($PSCmdlet.ParameterSetName -eq "UseDotnetCli") {
+        #perform dotnet pack using a temporary project file.
+        $dotnetCliPath = (Get-Command -Name "dotnet").Source
```
We could do for symmetry I guess, however NugetExePath is fed in as an input, the dotnet cli command is found using Get-Command and indexing into .Source on line 58 would throw an error which would be picked up by the try/catch in line 382 of the calling Publish-PSArtifactUtility function. Personally I prefer this approach and would rather have the nuget.exe case call get-command nuget.exe or something similar but I am not sure of all the ramifications and where else exactly nuget.exe is searched for and found using PowershellGet. Happy to change though unless you can shed any more light on it.

edyoung replied:
OK

jzabroski replied:
@Benny1007 @edyoung I would like a Verbose Preference to enable logging which Get-Command is retrieved.  What do you think?

Benny1007 commented:
Have addressed most feedback. I can't seem to get these tests to even complete locally or on the build servers though. Do you have any tips? - The test suite is pretty scary.

alerickson commented:
@Benny1007 check out https://github.com/PowerShell/PowerShellGet#running-tests for running tests locally.  You can also run Invoke-Pester with a particular test file instead of running the entire suite. 
Also be sure to first run "import-module src/PowerShellGet -force" to ensure you have your latest changes before running tests.

Benny1007 commented:
> @Benny1007 check out https://github.com/PowerShell/PowerShellGet#running-tests for running tests locally. You can also run Invoke-Pester with a particular test file instead of running the entire suite.
> Also be sure to first run "import-module src/PowerShellGet -force" to ensure you have your latest changes before running tests.

Thanks, so I have run the entire suite locally and everything passes. 
I have also run the entire suite on my macbook with the latest powershell, this seems to have other issues in some other tests related to permissions to temp drive. 

I have no idea why they are not working on the build server, but it's like they never even finish.

edyoung commented:
The tests fail because you're calling a function with a mandatory param but not passing a value. Powershell then stops everything and prompts for the user to supply the argument, which isn't so useful on a CI server, then times out waiting. See this snippet from Travis CI log:


cmdlet New-NuspecFile at command pipeline position 1

Supply values for the following parameters:

Id: 


No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Benny1007 commented:
Cool thanks. It's concerning none of this was picked up running the test suite locally. So within the calling function, we can't rely on $PSModuleInfo.Name or $PSScriptInfo.Name having a value or do we just need to fix up the test data where it creates these modules/scripts?

Benny1007 commented:
Something that has led to false positives during testing is that if I unload PowerShellGet module with:-
`Remove-Module -Name PowerShellGet -Force`
then load it from the code base with 
`Import-Module -Name .\src\PowerShellGet\PowerShellGet.psd1 -Force`
Then break on say line 119 of PSGet.PublishModule.Tests
I then seem to have 2 PowerShellGet modules loaded. I think this is caused by line 1160 in PSGetTestUtils.psm1
This caused the newer PowerShellGet from C:\Program Files\WindowsPowerShell\Modules\PowerShellGet version 2.1.2 to be loaded and used hence forth masking my issues. 
You come across that before?

alerickson commented:
@Benny1007 no I haven't.  Are you following option 1 or option 2 when running tests (https://github.com/PowerShell/PowerShellGet#running-tests)?  I've had issues with the env path unintentionally changing and importing the wrong module, so I prefer to follow the latter.

edyoung commented:
I always set $env:psmodulepath to pick up the version under development rather than the installed one, that avoids issues like that

Benny1007 commented:
As mentioned in the PR feedback truncating of the Tags has been removed and replace with a warning which won't exactly help #344 but will provided a clearer reason why the publish has failed?

awickham10 commented:
> As mentioned in the PR feedback truncating of the Tags has been removed and replace with a warning which won't exactly help #344 but will provided a clearer reason why the publish has failed?

It seems like the problem is Publish-Module includes every function as a tag. What if a switch was added to exclude functions as tags? Something like... `Publish-Module -Name 'MyModule' -ExcludeFunctionsAsTags` that way the odds of hitting the 4000 character limit is drastically reduced?

edyoung commented:
@awickham10 I'd be fine with that (or a -EnforceMaximumTagLength option maybe), but I'd rather keep that a separate change so this is a pure refactor

edyoung commented:
Thanks @Benny1007 !

alerickson commented:
The Visual Studio 2017, PowerShell Core environment on AppVeyor says it's passing but didn't actually run.  See: https://ci.appveyor.com/project/PowerShell/powershellget/builds/23226384/job/hu9xotrcnsvkyb8a

Benny1007 commented:
That's not very good. The build server reports success on a failure?
Looks like the build server didn't have PowerShell core (pwsh.exe) installed?

jzabroski commented on src/PowerShellGet/private/functions/New-NugetPackage.ps1:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    if ($PSCmdlet.ParameterSetName -eq "UseDotnetCli") {
+        #perform dotnet pack using a temporary project file.
```
@edyoung  @Benny1007  Extract this into it's own function.  Replace the comment with a function name.  Use a PowerShell approved verb.

Benny1007 replied:
what are the benefits of abstracting this further. Don't want to get into similar situation of Register-PSRepository and Resolve-Location, Get-LocationString, Get-EscapedString etc... This function has a single responsibility - to create a nuget package, having all the code at hand in one file to perform that was the reason for abstracting it in the first place. Personally not a fan of splitting out to multiple functions for the sake of it.

jzabroski replied:
It's not for the sake of it.  It's for building reusable abstractions.  Why is it out fault that the Process API is not designed for consuming a quick task's exit code and standard output and error streams?  We just wrap that and clearly express we care about those 3 things.  See my comment here which when I replied by email 2 days ago somehow got stuck at the bottom of the review: https://github.com/PowerShell/PowerShellGet/pull/444#issuecomment-477119239

jzabroski commented on src/PowerShellGet/private/functions/New-NugetPackage.ps1:
```diff
@@ -0,0 +1,110 @@
+
+    if ($PSCmdlet.ParameterSetName -eq "UseDotnetCli") {
+        #perform dotnet pack using a temporary project file.
+        $dotnetCliPath = (Get-Command -Name "dotnet").Source
```
Create Verbose Preference to log which Get-Command is retrieved.  This would be useful in submitting bug reports.  It could even verbose the version of the exe found.

jzabroski commented on src/PowerShellGet/private/functions/New-NugetPackage.ps1:
```diff
@@ -0,0 +1,110 @@
+                Remove-Item -Path $tempPath -Force -Recurse
+            }
+            $stdOut = $process.StandardOut.ReadToEnd()
+            throw "dotnet cli failed to pack $stdOut"
```
This still does not allow inspecting the temporary created project file.  This makes it a bit magical when trying to understand what got uploaded to a repository as part of the publish process.

@Benny1007 @edyoung

Benny1007 replied:
Overwrite the outputpath variable and then you would know where the artifact resides.

Benny1007 replied:
sorry, was missing a question mark off the last comment! would need to pass this through from the public function.

jzabroski commented:
@edyoung @Benny1007  I just saw this commit - for some reason it showed up in my GitHub Read folder instead of notifications.  I only reviewed one file so far, though.

alerickson commented:
@jzabroski just meant that the PR shouldn't have been merged because the tests for core didn't run.

jzabroski commented on src/PowerShellGet/private/functions/New-NuspecFile.ps1:
```diff
@@ -0,0 +1,125 @@
+
+    if ($Tags.Length -gt 4000) {
+        $Tags = $Tags.Substring(0, $Tags.LastIndexOf(" "))
+        Write-Warning -Message "Nuspec 'Tag' list exceeded max 4000 characters and was truncated."
```
This appears to address https://github.com/PowerShell/PowerShellGet/issues/344

Additionally, now that this is a first-class function exposable to end-users, the additional request to make 4000 limit configurable for users who want to upload packages to their own nuget server is possible

Benny1007 replied:
These nuget functions are still private functions and not exposed in the module to the public.

jzabroski replied:
Then how am I supposed to follow your advice to overwrite the output path?   See your comment here https://github.com/PowerShell/PowerShellGet/pull/444#discussion_r269376413 - maybe I'm missing something obvious and I apologize for being stubborn if that's the case.

Benny1007 replied:
There would have to be another separate PR that exposed the outputpath via the public members. This small refactor was just the start of the re-write of Publish-Module I assume the refactoring for this module will be across Publish-PSArtifactUtility (potentially going away) and Publish-Module.

jzabroski commented on src/PowerShellGet/private/functions/New-NuspecFile.ps1:
```diff
@@ -0,0 +1,125 @@
+    $packageElement = $xml.CreateElement("package", $nameSpaceUri)
+    $metaDataElement = $xml.CreateElement("metadata", $nameSpaceUri)
+
+    #truncate tags if they exceed nuspec specifications for size.
```
If you're destructively messing with parameters, create a Process block attached to the params block.

Benny1007 replied:
can you explain this further?

jzabroski replied:
In PowerShell v1, the only way to validate parameters was in a Process block and to throw an exception.  In PowerShell v2, you can use ValidateScript.

Benny1007 replied:
I think this comment was in reference to the truncation of tags?, which is now NOT happening but I seem to have left the comment in the code.

jzabroski commented on src/PowerShellGet/private/functions/New-NuspecFile.ps1:
```diff
@@ -0,0 +1,125 @@
+    $Tags = $Tags -Join " "
+
+    if ($Tags.Length -gt 4000) {
+        $Tags = $Tags.Substring(0, $Tags.LastIndexOf(" "))
```
This is not guaranteed to truncate at the lesser of 4,000 characters or "full list of tags separated by spaces up to 4,000 characters"

jzabroski replied:
@Benny1007  Here is a C# way to do what I'm describing. The author is kind enough to post sample data that you can use to test against.  I also recommend making this a private helper function for readability: https://codereview.stackexchange.com/a/163788/193241

Benny1007 replied:
The snippet is outdated, it was recommended in the PR feedback to NOT truncate.

jzabroski commented:
How am I supposed to overwrite the output path on a private function?

On Tue, Mar 26, 2019, 9:28 PM benny1007 <notifications@github.com> wrote:

> *@Benny1007* commented on this pull request.
> ------------------------------
>
> In src/PowerShellGet/private/functions/New-NuspecFile.ps1
> <https://github.com/PowerShell/PowerShellGet/pull/444#discussion_r269376643>
> :
>
> > +    $nameSpaceUri = "http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd"
> +    [xml]$xml = New-Object System.Xml.XmlDocument
> +
> +    $xmlDeclaration = $xml.CreateXmlDeclaration("1.0", "utf-8", $null)
> +    $xml.AppendChild($xmlDeclaration) | Out-Null
> +
> +    #create top-level elements
> +    $packageElement = $xml.CreateElement("package", $nameSpaceUri)
> +    $metaDataElement = $xml.CreateElement("metadata", $nameSpaceUri)
> +
> +    #truncate tags if they exceed nuspec specifications for size.
> +    $Tags = $Tags -Join " "
> +
> +    if ($Tags.Length -gt 4000) {
> +        $Tags = $Tags.Substring(0, $Tags.LastIndexOf(" "))
> +        Write-Warning -Message "Nuspec 'Tag' list exceeded max 4000 characters and was truncated."
>
> These nuget functions are still private functions and not exposed in the
> module to the public.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PowerShell/PowerShellGet/pull/444#discussion_r269376643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbT_Z0fGzjHdBtWmI7H9mTMHCgf4wAaks5vaskmgaJpZM4b0VCL>
> .
>

jzabroski commented:
What's the benefit of having nuget.exe instead of nugetupdate.exe,
nugetpublish.exe, nugethelp.exe? Reductio ad absurdum.

If you think my suggestion makes the code less readable, that's OK. But I
actually think it would improve readability since all you really care about
is "do something, aggregate stdout, srderr and exit code, and return." I
think the fact it's already not handled consistently across the code base
is reason enough to add a helper. Call it Helper/Invoke-Process so that
it's clearer its a cross cutting concept. At the end of the day you should
not care about the ceremony of Process start, redirection and exit codes.



On Tue, Mar 26, 2019, 9:24 PM benny1007 <notifications@github.com> wrote:

> *@Benny1007* commented on this pull request.
> ------------------------------
>
> In src/PowerShellGet/private/functions/New-NugetPackage.ps1
> <https://github.com/PowerShell/PowerShellGet/pull/444#discussion_r269376094>
> :
>
> > +        $processStartInfo.UseShellExecute = $false
> +        $processStartInfo.Arguments = $ArgumentList
> +
> +        $process = New-Object System.Diagnostics.Process
> +        $process.StartInfo = $processStartInfo
> +        $process.Start() | Out-Null
> +        $process.WaitForExit()
> +
> +        if (-Not ($process.ExitCode -eq 0 )) {
> +            $stdErr = $process.StandardError.ReadToEnd()
> +            throw "nuget.exe failed to pack $stdErr"
> +        }
> +    }
> +
> +    if ($PSCmdlet.ParameterSetName -eq "UseDotnetCli") {
> +        #perform dotnet pack using a temporary project file.
>
> what are the benefits of abstracting this further. Don't want to get into
> similar situation of Register-PSRepository and Resolve-Location,
> Get-LocationString, Get-EscapedString etc... This function has a single
> responsibility - to create a nuget package, having all the code at hand in
> one file to perform that was the reason for abstracting it in the first
> place. Personally not a fan of splitting out to multiple functions for the
> sake of it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PowerShell/PowerShellGet/pull/444#discussion_r269376094>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbT_eyRx9X1Qi9IScNxQC6vfG2oP5tYks5vashagaJpZM4b0VCL>
> .
>

jzabroski commented:
> @awickham10 I'd be fine with that (or a -EnforceMaximumTagLength option maybe), but I'd rather keep that a separate change so this is a pure refactor

This is a private function right now so it doesn't add a lot of value for something that will never have code coverage in production usage.  The more important question is whether some of these helper functions should be public instead of private.  My gut feeling is that they should be split out into a separate helper module, so that if people want to debug PowerShellGet internals, they can.  This is not a traditional way of thinking about using modules, but it just "Feels right" to me.  Also, by making it a separate module instead of public interface to a major mainstream package, it allows for easier user story to sunset the module once PowerShellGet becomes more stable over time and the need for these utility functions goes away from the public eye.
